### PR TITLE
TUITextRenderer auto-scroll.

### PR DESCRIPTION
**Pull request by avaidyam**
If a `TUITextRenderer` is within a `TUIScrollView` it now auto-scrolls with keybindings. There are a few new keybindings I had to add to accomplish this, although they'd be better placed in the other PR. Upon merge, they should merge fine.

### CODE REVIEW

**Comment by jspahrsummers** [comment]
Some of this code is duplicated from another one of your pull requests. I don't want to duplicate the work of reviewing it, so this should wait until the other one has been merged in.

Please note such dependencies in the description of your PRs, if possible.

**Comment by avaidyam** [comment]
The dependencies have been met, and I believe @jspahrsummers can begin reviewing this PR.

**Comment by jspahrsummers** [comment]
Please update this branch with master so that any duplicate code is removed from the diff.

**Comment by avaidyam** [comment]
@jspahrsummers @jwilling This is now up to date and tabbed properly.

File: lib/UIKit/TUITextEditor.m [hunk 1]
```diff
@@ -146,6 +150,8 @@ - (void)paste:(id)sender
 	if (self.editable) {
 		[self insertText:[[NSPasteboard generalPasteboard] stringForType:NSPasteboardTypeString]];
 	}
+	[self insertText:[[NSPasteboard generalPasteboard] stringForType:NSPasteboardTypeString]];
```
Comment (jspahrsummers) [inline lib/UIKit/TUITextEditor.m]: This will insert the pasteboard text twice.

File: lib/UIKit/TUITextRenderer+KeyBindings.m [hunk 2]
```diff
@@ -74,6 +77,21 @@ - (TUITextEditor *)_textEditor
 	return nil;
 }
 
+- (void)_scrollToIndex:(long)index {
```
Comment (jspahrsummers) [inline lib/UIKit/TUITextRenderer+KeyBindings.m]: This should be declared in `TUITextRenderer+Private` so the implementation doesn't have to declare it.

This method is always used in the same way – to scroll to the selection. Why not just make it into a method for doing exactly that?

File: lib/UIKit/TUITextRenderer+KeyBindings.m [hunk 3]
```diff
@@ -359,5 +399,33 @@ - (void)moveToEndOfDocument:(id)sender
...
+{
+	[self deleteToBeginningOfLine:sender];
+}
+
+- (void)deleteToEndOfLine:(id)sender
+{
+	NSRange deleteRange = [self selectedRange];
+	if(deleteRange.length == 0)
+		deleteRange.length = [TEXT length] - deleteRange.location;
+	killBuffer = [[self _textEditor].backingStore attributedSubstringFromRange:deleteRange];
```
Comment (jspahrsummers) [inline lib/UIKit/TUITextRenderer+KeyBindings.m:423]: This needs to be set for basically every deletion method.

File: lib/UIKit/TUITextRenderer+KeyBindings.m [hunk 4]
```diff
@@ -74,6 +77,21 @@ - (TUITextEditor *)_textEditor
 	return nil;
 }
 
+- (void)_scrollToIndex:(long)index {
+	if(self.view && [self.view.superview isKindOfClass:[TUIScrollView class]]) {
```
Comment (jspahrsummers) [inline lib/UIKit/TUITextRenderer+KeyBindings.m]: :-1: This should use a pattern similar to `NSView.enclosingScrollView` – we should never depend on a certain view hierarchy setup.

File: lib/UIKit/TUITextRenderer+KeyBindings.m [hunk 5]
```diff
@@ -74,6 +77,21 @@ - (TUITextEditor *)_textEditor
...
+			TUIScrollView *scrollView = (TUIScrollView*)self.view.superview;
+			CFRange r = { index, 0 };
```
Comment (jspahrsummers) [inline lib/UIKit/TUITextRenderer+KeyBindings.m]: Use `CFRangeMake` instead.

File: lib/UIKit/TUITextRenderer+KeyBindings.m [hunk 6]
```diff
@@ -74,6 +77,21 @@ - (TUITextEditor *)_textEditor
...
+			CFIndex nRects = 1;
+			CGRect rects[nRects];
+			AB_CTFrameGetRectsForRange([self ctFrame], r, rects, &nRects);
+			if(nRects == 1)
+				[scrollView scrollRectToVisible:rects[0] animated:YES];
+			else if(nRects == 0)
+				[scrollView scrollRectToVisible:CGRectMake(0, self.view.frame.size.height / 2, 0, 1)
```
Comment (jspahrsummers) [inline lib/UIKit/TUITextRenderer+KeyBindings.m]: It's not clear to me what this is trying to accomplish.

**Comment by jspahrsummers** [comment]
Please make this match the [style guide](https://github.com/github/objective-c-conventions).

File: lib/UIKit/TUITextEditor.m (hunk 1 above)
Comment (avaidyam) [inline lib/UIKit/TUITextEditor.m]: Merge issue!

File: lib/UIKit/TUITextRenderer+KeyBindings.m (hunk 4 above)
Comment (avaidyam) [inline lib/UIKit/TUITextRenderer+KeyBindings.m]: How would I set that up? EDIT: Figured it out!

File: lib/UIKit/TUITextRenderer+KeyBindings.m (hunk 2 above)
Comment (avaidyam) [inline lib/UIKit/TUITextRenderer+KeyBindings.m]: Sorry, I didn't quite understand what you meant. Would you please elaborate?

File: lib/UIKit/TUITextRenderer+KeyBindings.m (hunk 6 above)
Comment (avaidyam) [inline lib/UIKit/TUITextRenderer+KeyBindings.m]: This gets the scroll view, if there is one, then, if there are indeed rects for the index provided, it scrolls upwards, otherwise, it scrolls to the beginning of the text if there are none for the index given.
This `long` should also be a `CFIndex` and I'll make that change.

File: lib/UIKit/TUITextRenderer+KeyBindings.m (hunk 2 above)
Comment (avaidyam) [inline lib/UIKit/TUITextRenderer+KeyBindings.m]: In regards to declaration, it's used solely by the keybindings category, so it should be fine, undeclared.

File: lib/UIKit/TUITextRenderer+KeyBindings.m [hunk 7]
```diff
@@ -74,6 +77,27 @@ - (TUITextEditor *)_textEditor
...
+		
+		// If it exists, then scroll the the beginning of the rects.
+		if(nRects == 1)
+			[scrollView scrollRectToVisible:rects[0] animated:YES];
+		
+		// Otherwise, scroll to the middle of the text rects.
+		else if(nRects == 0)
+			[scrollView scrollRectToVisible:CGRectMake(0, self.view.frame.size.height / 2, 0, 1)
+								   animated:YES];
+		}
```
Comment (jspahrsummers) [inline lib/UIKit/TUITextRenderer+KeyBindings.m]: Incorrect indentation here.

File: lib/UIKit/TUIView.m [hunk 8]
```diff
@@ -975,6 +976,12 @@ - (TUIView *)firstSuperviewOfClass:(Class)c
 	return [self.superview firstSuperviewOfClass:c];
 }
 
+- (TUIScrollView *)enclosingScrollView {
+	if([self.superview isKindOfClass:TUIScrollView.class])
+		return (TUIScrollView *)self.superview;
+	else return nil;
```
Comment (jspahrsummers) [inline lib/UIKit/TUIView.m]: This isn't a correct implementation of this method. A correct implementation would walk all superviews until a scroll view is found.

Furthermore, this is already implemented as `ancestorScrollView`. See `<TUIBridgedView>`.

**Comment by jspahrsummers** [comment]
:mailbox:

File: lib/UIKit/TUIView.m (hunk 8 above)
Comment (avaidyam) [inline lib/UIKit/TUIView.m]: But then the whole auto-scroll wouldn't work if it was within, say a
`TUITableViewCell` and it tried to shift the whole table view.

**Comment by avaidyam** [comment]
@jspahrsummers This is now fixed.

**Comment by jspahrsummers** [comment]
@galaxas0 It's not. I mentioned `ancestorScrollView` – that should be used instead.

**Comment by avaidyam** [comment]
Ah, alright. I've got it now.

**Comment by avaidyam** [comment]
@jspahrsummers :mailbox_with_mail: This was fine for some time now.

File: lib/UIKit/TUITextRenderer+KeyBindings.m [hunk 9]
```diff
@@ -74,6 +79,27 @@ - (TUITextEditor *)_textEditor
 	return nil;
 }
 
+- (void)_scrollToIndex:(CFIndex)index {
+	TUIScrollView *scrollView = (TUIScrollView *)self.view.ancestorScrollView;
```
Comment (jspahrsummers) [inline lib/UIKit/TUITextRenderer+KeyBindings.m]: This should use the `<TUIBridgedScrollView>` interface. It's unsafe to assume that the `ancestorScrollView` is actually a `TUIScrollView` (as opposed to, e.g., an `NSClipView`).

**Comment by jspahrsummers** [comment]
Sorry, I didn't see a notification for this after your comment (which was made before the commit was pushed).

Got one more note.

**Comment by avaidyam** [comment]
Alright, should be good to go now! Thanks!

File: lib/UIKit/TUITextRenderer+KeyBindings.m [hunk 10]
```diff
@@ -74,6 +79,27 @@ - (TUITextEditor *)_textEditor
 	return nil;
 }
 
+- (void)_scrollToIndex:(CFIndex)index {
+	TUIScrollView <TUIBridgedScrollView> *scrollView = (TUIScrollView <TUIBridgedScrollView> *)self.view.ancestorScrollView;
```
Comment (jspahrsummers) [inline lib/UIKit/TUITextRenderer+KeyBindings.m]: This doesn't fix what I was describing. This variable should be of type `id<TUIBridgedScrollView>`, and no `TUIScrollView`-specific methods should be invoked upon it.

**Comment by avaidyam** [comment]
@jspahrsummers :mailbox_with_mail:

**Comment by avaidyam** [comment]
@jspahrsummers :mailbox_with_mail: I've fixed the issue here.

File: lib/UIKit/TUITextRenderer+KeyBindings.m [hunk 11]
```diff
@@ -74,6 +79,26 @@ - (TUITextEditor *)_textEditor
 	return nil;
 }
 
+- (void)_scrollToIndex:(CFIndex)index {
+	id <TUIBridgedScrollView> scrollView = (id <TUIBridgedScrollView>)self.view.ancestorScrollView;
```
Comment (jspahrsummers) [inline lib/UIKit/TUITextRenderer+KeyBindings.m]: This doesn't need a cast.

File: lib/UIKit/TUIView.h [hunk 12]
```diff
@@ -276,6 +276,8 @@ enum {
 
 @end
 
+@class TUIScrollView;
```
Comment (jspahrsummers) [inline lib/UIKit/TUIView.h]: Why is this here?

File: lib/UIKit/TUIView.m [hunk 13]
```diff
@@ -14,15 +14,16 @@
...
 #import <pthread.h>
 #import "NSColor+TUIExtensions.h"
 #import "TUICGAdditions.h"
+#import "TUIView.h"
 #import "TUILayoutManager.h"
 #import "TUINSView.h"
 #import "TUINSWindow.h"
 #import "TUITextRenderer.h"
 #import "TUIViewController.h"
+#import "TUIScrollView.h"
```
Comment (jspahrsummers) [inline lib/UIKit/TUIView.m]: Why is this here?

File: lib/UIKit/TUIView.h (hunk 12 above)
Comment (avaidyam) [inline lib/UIKit/TUIView.h]: Oops, I'll remove it right away. That was from `enclosingScrollView`.

**Comment by avaidyam** [comment]
@jspahrsummers :mailbox_closed: Patched those up too.

**Comment by jspahrsummers** [comment]
Thanks!
